actor-typed/src/main/scala/org/apache/pekko/actor/typed/internal/Supervision.scala

```diff
@@ -398,8 +404,10 @@ private class RestartSupervisor[T, Thr <: Throwable: ClassTag](initial: Behavior
       val newBehavior = Behavior.validateAsInitial(Behavior.start(initial, ctx.asInstanceOf[TypedActorContext[T]]))
       val nextBehavior = restartingInProgress match {
         case OptionVal.Some((stashBuffer, _)) =>
+          val behavior = stashBuffer.unstashAll(newBehavior.unsafeCast)
+          // reset stash state if not more exception throw
           restartingInProgress = OptionVal.None
-          stashBuffer.unstashAll(newBehavior.unsafeCast)
+          behavior
```

Comment:
Only reset if the unstashAll processed success, otherwise the unprocessed messages will be lost right?

Reply:
Yes, because unhandled messages are stored in the StashBuffer within the `restartingInProgress` field.